File: apps/api/app/schemas/learn.py
"""Pydantic schemas for learning tracks."""
from __future__ import annotations

from typing import Literal

from pydantic import Field, HttpUrl

from .base import CamelModel


class ResourceRead(CamelModel):
    title: str
    url: HttpUrl
    type: Literal["article", "video", "docs", "github", "exercise"]
    description: str | None = None


class ModuleRead(CamelModel):
    title: str
    objective: str
    resources: list[ResourceRead] = Field(min_length=1, max_length=5)
    task: str


class CapstoneRead(CamelModel):
    title: str
    brief: str
    deliverable: str


class BadgeRead(CamelModel):
    name: str
    description: str
    icon: str | None = None


class LearningTrackRead(CamelModel):
    id: str
    title: str
    level: Literal["beginner", "intermediate", "advanced"]
    time_estimate_hours: int = Field(gt=0)
    outcome: str
    modules: list[ModuleRead] = Field(min_length=3)
    capstone: CapstoneRead
    badge: BadgeRead

File: packages/types/src/schemas/learn.ts
import { z } from "zod";

export const learnResourceSchema = z.object({
  title: z.string(),
  url: z.string().url(),
  type: z.enum(["article", "video", "docs", "github", "exercise"]),
  description: z.string().optional(),
});

export const learnModuleSchema = z.object({
  title: z.string(),
  objective: z.string(),
  resources: z.array(learnResourceSchema).min(1).max(5),
  task: z.string(),
});

export const capstoneSchema = z.object({
  title: z.string(),
  brief: z.string(),
  deliverable: z.string(),
});

export const badgeSchema = z.object({
  name: z.string(),
  description: z.string(),
  icon: z.string().optional(),
});

export const learnTrackSchema = z.object({
  id: z.string(),
  title: z.string(),
  level: z.enum(["beginner", "intermediate", "advanced"]),
  timeEstimateHours: z.number().int().positive(),
  outcome: z.string(),
  modules: z.array(learnModuleSchema).min(3),
  capstone: capstoneSchema,
  badge: badgeSchema,
});

export type LearnResource = z.infer<typeof learnResourceSchema>;
export type LearnModule = z.infer<typeof learnModuleSchema>;
export type LearnCapstone = z.infer<typeof capstoneSchema>;
export type LearnBadge = z.infer<typeof badgeSchema>;
export type LearnTrack = z.infer<typeof learnTrackSchema>;
